src/cpp/fastdds/publisher/DataWriterImpl.cpp

```diff
@@ -676,22 +676,27 @@ void DataWriterImpl::InnerDataWriterListener::onWriterMatched(
         data_writer_->listener_->on_publication_matched(
             data_writer_->user_datawriter_, info);
     }
-
-    data_writer_->publisher_->publisher_listener_.on_publication_matched(data_writer_->user_datawriter_, info);
+    else
+    {
+        data_writer_->publisher_->publisher_listener_.on_publication_matched(data_writer_->user_datawriter_, info);
+    }
 }
 
 void DataWriterImpl::InnerDataWriterListener::on_offered_incompatible_qos(
         RTPSWriter* /*writer*/,
         fastdds::dds::PolicyMask qos)
 {
     OfferedIncompatibleQosStatus& status = data_writer_->update_offered_incompatible_qos(qos);
-    if (data_writer_->listener_ != nullptr)
+    if (data_writer_->listener_ != nullptr  &&
+            (data_writer_->user_datawriter_->get_status_mask() & StatusMask::offered_incompatible_qos()).any())
     {
         data_writer_->listener_->on_offered_incompatible_qos(data_writer_->user_datawriter_, status);
         status.total_count_change = 0u;
     }
-
-    data_writer_->publisher_->publisher_listener_.on_offered_incompatible_qos(data_writer_->user_datawriter_, status);
+    else
+    {
+        data_writer_->publisher_->publisher_listener_.on_offered_incompatible_qos(data_writer_->user_datawriter_, status);
```

Comment:
This is the inner listener of the Publisher. It is this callback the one that should check the user's mask (and the existence of the user's listener)

Which currently is not doing, because of the wrong implementation of the listener logic. That was also corrected before I reverted that commit. The thing is it won't be completely correct if we don't correct the DomainParticipant's inner listener too. Don't know if it is worth changing all that on this single status, having all the rest wrong. At least, they are consistently wrong now.